docs/PROTOCOL.md

```diff
@@ -0,0 +1,243 @@
+    abi.encode(
+        ch.participants,
+        ch.adjudicator,
+        ch.challenge,
+        ch.nonce
+    )
+);
+```
+
+### 4.2. State Hash
+
+The stateHash MUST be computed as:
+
+```solidity
+stateHash = keccak256(
+    abi.encode(
+        channelId,
+        state.data,
+        state.allocations
+    )
+);
+```
+
+For signature verification, the final hash MUST use Ethereum's personal sign message format:
+```solidity
+signatureHash = keccak256(
+    abi.encodePacked(
+        "\x19Ethereum Signed Message:\n32",
+        stateHash
+    )
+);
+```
+
+## 5. Funding Protocol
+
+### 5.1. Creation Phase
+
+1. The Creator MUST:
+   - Construct a channel configuration with participants, adjudicator, challenge period, and nonce
+   - Prepare an initial state where `state.data` MUST be set to the magic number `CHANOPEN` (7877)
+   - Define expected token deposits for all participants in the `state.allocations` array
+   - Sign the stateHash of this initial state
+   - Include their signature in the `state.sigs` array at position 0
+   - Call the `create` function with the channel configuration and initial state
+
+2. The system MUST:
+   - Verify the Creator's signature on the funding stateHash
+   - Lock the Creator's funds according to the allocation
+   - Set the channel status to `INITIAL`
+   - Emit a `Created` event with the channelId, channel configuration, and expected deposits
+
+### 5.2. Joining Phase
+
+1. Each non-Creator participant MUST:
+   - Verify the channelId and expected allocations
+   - Sign the same funding stateHash (containing the magic number `CHANOPEN`)
+   - Call the `join` function with the channelId, their participant index, and signature
+
+2. The system MUST:
+   - Verify the participant's signature against the funding stateHash
+   - Confirm the signer matches the expected participant at the given index
+   - Lock the participant's funds according to the allocation
+   - Track the actual deposit in the channel metadata
+   - Emit a `Joined` event with the channelId and participant index
+
+3. When all participants have joined, the system MUST:
+   - Verify that all expected deposits are fulfilled
+   - Set the channel status to `ACTIVE`
+   - Emit an `Opened` event with the channelId
+
+## 6. Off-Chain State Updates
+
+1. Participants MUST:
+   - Exchange and sign state updates off-chain
+   - Maintain a record of the latest valid state
+   - Use application-specific data in the `state.data` field
+
+2. Each new state MUST:
+   - Contain updated allocations reflecting the new distribution of assets
+   - Be signed by the necessary participants according to application rules
+   - Comply with the validation rules of the channel's adjudicator
+
+## 7. Channel Closure
+
+### 7.1. Cooperative Close
+
+1. To close cooperatively, any participant MAY:
+   - Prepare a final state where `state.data` MUST be set to the magic number `CHANCLOSE` (7879)
+   - Collect signatures from all participants on this final state
```

Comment:
Maybe as an idea, we can also make a close method similar to an open:
1. Initiator of the close submit close request
2. Another participant should receive the needed notification
3. Approve the close request and conclude the channel, or challenge it

And as a security mechanism, add a cooldown mechanism for expired close requests

Reply:
I think it is better to have "close" atomic to avoid dealing with half-transitioned state.

Reply:
One-sided close with approve or disapprove duration is already implemented (designed) it's called "challenge"

Reply:
ok ok, make sense I guess

Reply:
I thought about this too @mnaichuk we may fallback to it, if we don't find an easy way to happy-close